jpyinterpreter/src/main/java/org/optaplanner/jpyinterpreter/opcodes/variable/LoadGlobalOpcode.java

```diff
@@ -17,26 +19,63 @@ public LoadGlobalOpcode(PythonBytecodeInstruction instruction) {
         super(instruction);
     }
 
+    private int getGlobalIndex(FunctionMetadata functionMetadata) {
+        return (functionMetadata.pythonCompiledFunction.pythonVersion.compareTo(PythonVersion.PYTHON_3_11) >= 0)
+                ? instruction.arg >> 1
+                : instruction.arg;
+    }
+
+    private boolean pushNullBeforeGlobal(FunctionMetadata functionMetadata) {
+        return functionMetadata.pythonCompiledFunction.pythonVersion.compareTo(PythonVersion.PYTHON_3_11) >= 0
+                && ((instruction.arg & 1) == 1);
+    }
+
+    private PythonLikeObject getGlobal(FunctionMetadata functionMetadata) {
+        return functionMetadata.pythonCompiledFunction.globalsMap
+                .get(functionMetadata.pythonCompiledFunction.co_names.get(getGlobalIndex(functionMetadata)));
+    }
+
     @Override
     protected StackMetadata getStackMetadataAfterInstruction(FunctionMetadata functionMetadata, StackMetadata stackMetadata) {
-        PythonLikeObject global = functionMetadata.pythonCompiledFunction.globalsMap
-                .get(functionMetadata.pythonCompiledFunction.co_names.get(instruction.arg));
-        if (global != null) {
-            return stackMetadata.push(ValueSourceInfo.of(this, global.__getType()));
+        boolean pushNull = pushNullBeforeGlobal(functionMetadata);
+        PythonLikeObject global = getGlobal(functionMetadata);
+
+        if (pushNull) {
+            if (global != null) {
+                return stackMetadata
+                        .push(ValueSourceInfo.of(this, BuiltinTypes.BASE_TYPE))
+                        .push(ValueSourceInfo.of(this, global.__getType()));
+            } else {
+                return stackMetadata
+                        .push(ValueSourceInfo.of(this, BuiltinTypes.BASE_TYPE))
+                        .push(ValueSourceInfo.of(this, BuiltinTypes.BASE_TYPE));
+            }
         } else {
-            return stackMetadata.push(ValueSourceInfo.of(this, BuiltinTypes.BASE_TYPE));
+            if (global != null) {
+                return stackMetadata
+                        .push(ValueSourceInfo.of(this, global.__getType()));
+            } else {
+                return stackMetadata
+                        .push(ValueSourceInfo.of(this, BuiltinTypes.BASE_TYPE));
+            }
         }
     }
 
     @Override
     public void implement(FunctionMetadata functionMetadata, StackMetadata stackMetadata) {
-        PythonLikeObject global = functionMetadata.pythonCompiledFunction.globalsMap.get(
-                functionMetadata.pythonCompiledFunction.co_names.get(instruction.arg));
+        int globalIndex = getGlobalIndex(functionMetadata);
+        boolean pushNull = pushNullBeforeGlobal(functionMetadata);
+
+        PythonLikeObject global = getGlobal(functionMetadata);
+
         if (global instanceof CPythonType || global instanceof PythonObjectWrapper) {
             // TODO: note native objects are used somewhere
```

Comment:
Another TODO left behind?

Reply:
This TODO is for a future work, where errors, warnings, etc. are captured into a CompilationResult. For this TODO in particular, if native objects are used, compilation will still succeed, but the code will probably be slower than just calling the CPython method, since it is likely you need to eat the overhead cost for transferring from Java to CPython back at least once (and maybe multiple times).